tuf/ngclient/_internal/metadata_bundle.py

```diff
@@ -0,0 +1,461 @@
+        return self._bundle[role]
+
+    def __len__(self) -> int:
+        """Returns number of Metadata objects in bundle"""
+        return len(self._bundle)
+
+    def __iter__(self) -> Iterator[Metadata]:
+        """Returns iterator over all Metadata objects in bundle"""
+        return iter(self._bundle)
+
+    # Helper properties for top level metadata
+    @property
+    def root(self) -> Optional[Metadata]:
+        """Current root Metadata or None"""
+        return self._bundle.get("root")
+
+    @property
+    def timestamp(self) -> Optional[Metadata]:
+        """Current timestamp Metadata or None"""
+        return self._bundle.get("timestamp")
+
+    @property
+    def snapshot(self) -> Optional[Metadata]:
+        """Current snapshot Metadata or None"""
+        return self._bundle.get("snapshot")
+
+    @property
+    def targets(self) -> Optional[Metadata]:
+        """Current targets Metadata or None"""
+        return self._bundle.get("targets")
+
+    # Methods for updating metadata
+    def update_root(self, data: bytes):
+        """Verifies and loads 'data' as new root metadata.
+
+        Note that an expired intermediate root is considered valid: expiry is
+        only checked for the final root in root_update_finished().
+
+        Args:
+            data: unverified new root metadata as bytes
+
+        Raises:
+            RepositoryError: Metadata failed to load or verify. The actual
+                error type and content will contain more details.
+        """
+        if self._root_update_finished:
+            raise RuntimeError(
+                "Cannot update root after root update is finished"
+            )
+        logger.debug("Updating root")
+
+        try:
+            new_root = Metadata.from_bytes(data)
+        except DeserializationError as e:
+            raise exceptions.RepositoryError("Failed to load root") from e
+
+        if new_root.signed.type != "root":
+            raise exceptions.RepositoryError(
+                f"Expected 'root', got '{new_root.signed.type}'"
+            )
+
+        if self.root is not None:
+            # We are not loading initial trusted root: verify the new one
+            if not verify_with_threshold(self.root, "root", new_root):
+                raise exceptions.UnsignedMetadataError(
+                    "New root is not signed by root", new_root.signed
+                )
+
+            if new_root.signed.version != self.root.signed.version + 1:
+                raise exceptions.ReplayedMetadataError(
+                    "root", new_root.signed.version, self.root.signed.version
+                )
+
+        if not verify_with_threshold(new_root, "root", new_root):
+            raise exceptions.UnsignedMetadataError(
+                "New root is not signed by itself", new_root.signed
+            )
+
+        self._bundle["root"] = new_root
+        logger.debug("Updated root")
+
+    def root_update_finished(self):
```

Comment:
Where do you [check](https://github.com/trishankatdatadog/tuf-on-a-plane/blob/ee8049a63ea439d72509f6e876eef08fb50c6ab1/src/tuf_on_a_plane/repository.py#L209) for revoking timestamp/snapshot to recover from fast-forward attacks?

Reply:
We don't: 
* because we are strict about the order of loading different metadata, we cannot ever have timestamp or snapshot loaded at this point
* after this point, we cannot load local timestamp or snapshot if they are no longer signed by the correct (rotated) keys 
so I believe there's just nothing to revoke

The spec does also say _delete the trusted timestamp and snapshot metadata files_: we don't do that either currently:
* the request is a bit odd: the metadata files _by definition_ are not trusted anymore: the keys have been rotated
* There's just no way to load those files into the bundle anymore
We could do that... and if we don't we should definitely document why. I've been reluctant to add file removing unless really needed because it feels like removing evidence just as something maybe going wrong

Something else we don't make clear is the difference to current implementation: we have no "current" and "previous" metadata: Currently we store a single copy of each roles metadata (new metadata gets written as soon as it's validated by the bundle).

Reply:
> We could add the file removal of course... and if we don't we should definitely document why.

This is a good comment, we need some action to come from this. Even if it's just an issue capturing this intent.